From e865b52379b48c5ffb2577ea06d78e3f219670b7 Mon Sep 17 00:00:00 2001
From: Sasi Kumar Maddineni <quic_sasikuma@quicinc.com>
Date: Fri, 6 Jun 2025 09:09:22 +0530
Subject: [PATCH 1/2] mount: Allow building when macro MOUNT_ATTR_IDMAP is not
 available

This is to allow building the software on machines not having the
MOUNT_ATTR_IDMAP macro in header "linux/mount.h". When that macro is not
available, the dependency on struct mount_attr is also eliminated (which
is good since both the macro and the struct were added to the kernel
uapi virtually at the same time).

With the changes in this commit, errors would be thrown at runtime when
mounting the erofs image, but only if the idmap feature is used; this
resembles the behavior when the "new mount API" is not detected.

Upstream-Status: Accepted
[https://github.com/containers/composefs/pull/253]

Signed-off-by: Rogerio Guerra Borin <rogerio.borin@toradex.com>
Git-Commit: f163eba6ce8826007b3a1c10eb47fa41f490e217
Git-repo: https://github.com/composefs/composefs.git

[Sasi Kumar Maddineni <quic_sasikuma@quicinc.com>: Resolved merge conflicts]
Signed-off-by: Sasi Kumar Maddineni <quic_sasikuma@quicinc.com>
---
 composefs/libcomposefs/lcfs-mount.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/composefs/libcomposefs/lcfs-mount.c b/composefs/libcomposefs/lcfs-mount.c
index 3dd518b..4d1a729 100644
--- a/composefs/libcomposefs/lcfs-mount.c
+++ b/composefs/libcomposefs/lcfs-mount.c
@@ -108,6 +108,7 @@ static int syscall_move_mount(int from_dfd, const char *from_pathname, int to_df
 #endif
 }
 
+#ifdef HAVE_MOUNT_ATTR_IDMAP
 static int syscall_mount_setattr(int dfd, const char *path, unsigned int flags,
 				 struct mount_attr *attr, size_t usize)
 {
@@ -122,6 +123,7 @@ static int syscall_mount_setattr(int dfd, const char *path, unsigned int flags,
 	return -1;
 #endif
 }
+#endif
 
 #define MAX_DIGEST_SIZE 64
 
@@ -520,6 +522,7 @@ static errint_t lcfs_mount_erofs(const char *source, const char *target,
 		return -errno;
 
 	if (use_idmap) {
+#ifdef HAVE_MOUNT_ATTR_IDMAP
 		struct mount_attr attr = {
 			.attr_set = MOUNT_ATTR_IDMAP,
 			.userns_fd = state->options->idmap_fd,
@@ -529,6 +532,9 @@ static errint_t lcfs_mount_erofs(const char *source, const char *target,
 					    sizeof(struct mount_attr));
 		if (res < 0)
 			return -errno;
+#else
+                return -ENOTSUP;
+#endif
 	}
 
 	res = syscall_move_mount(fd_mnt, "", AT_FDCWD, target,
-- 
2.34.1

